### PR TITLE
Various RBAC API and CLI improvements (and some bonus bug fixes)

### PR DESCRIPTION
**Pull request by Kami**
This pull request includes various RBAC API and CLI improvements (and also some openapi related fixes):

* Fix ``system`` query param filter in ``/v1/rbac/roles`` API endpoint so it works correctly
* Update ``/v1/rbac/roles`` API endpoint so includes corresponding permission grant objects (see below for example). Previously we only included raw IDs so the response wasn't all that useful.
* Add new CLI commands for viewing RBAC roles:
    * ``st2 role list [--system]``
    * ``st2 role get <role id or name>``
* Add CLI commands for viewing RBAC user role assignments:
    * ``st2 role-assignment list [--role=<role name>] [--user=<username>]``
    * ``st2 role-assignment get <role assignment id>``

###  ``/v1/roles/`` API endpoint

Example output of ``/v1/roles/`` API endpoint before the above mentioned change:

```json
curl http://127.0.0.1:9101/v1/rbac/roles/role_timer
{
    "name": "role_timer",
    "permission_grants": [
        "58f4e3ad0640fd2a7948631f",
        "58f4e3ad0640fd2a79486320"
    ],
    "system": false,
    "id": "58f4e3ad0640fd2a7948631e",
    "description": "Role which grants all the action permissions on pack config_demo"
}
```

And after the change:

```json
curl http://127.0.0.1:9101/v1/rbac/roles/role_timer
{
    "name": "role_timer",
    "permission_grants": [
        "58f4e3ad0640fd2a7948631f",
        "58f4e3ad0640fd2a79486320"
    ],
    "system": false,
    "permission_grant_objects": [
        {
            "permission_types": [
                "trigger_list"
            ],
            "id": "58f4e3ad0640fd2a7948631f"
        },
        {
            "permission_types": [
                "trigger_view"
            ],
            "id": "58f4e3ad0640fd2a79486320",
            "resource_type": "trigger",
            "resource_uid": "trigger:core:c625e485-ceb8-45c6-8c7e-53577e7007cf:889c1c0b04f53100997af85f35ad3979"
        }
    ],
    "permission_grant_ids": [
        "58f4e3ad0640fd2a7948631f",
        "58f4e3ad0640fd2a79486320"
    ],
    "id": "58f4e3ad0640fd2a7948631e",
    "description": "Role which grants all the action permissions on pack config_demo"
}
```

Note: ``permission_grants`` field is duplicated and is there for backward compatibility reasons.

### st2 role CLI commands

```bash
st2 role list
2017-05-08 12:06:10,511  WARNING - Auth API server is not available, skipping authentication.
+--------------------------+--------------+--------+------------------------------------------+
| id                       | name         | system | description                              |
+--------------------------+--------------+--------+------------------------------------------+
| 58f4decf0640fd24a8fde770 | admin        | True   | admin                                    |
| 58f4decf0640fd24a8fde771 | observer     | True   | observer                                 |
| 58f4e3ad0640fd2a7948632a | role_one     | False  | Role which grants all the action         |
|                          |              |        | permissions on pack config_demo          |
| 58f4e3ad0640fd2a79486324 | role_three   | False  | Role which grants all the action         |
|                          |              |        | permissions on pack config_demo          |
| 58f4e3ad0640fd2a7948631e | role_timer   | False  | Role which grants all the action         |
|                          |              |        | permissions on pack config_demo          |
| 58f4e3ad0640fd2a79486321 | role_two     | False  | Role which grants all the action         |
|                          |              |        | permissions on pack config_demo          |
| 58f4e3ad0640fd2a79486327 | st2_admin    | False  | ST2 Admin Role                           |
| 58f4decf0640fd24a8fde772 | system_admin | True   | system_admin                             |
+--------------------------+--------------+--------+------------------------------------------+
```

```bash
st2 role list --system
+--------------------------+--------------+--------+--------------+
| id                       | name         | system | description  |
+--------------------------+--------------+--------+--------------+
| 58f4decf0640fd24a8fde770 | admin        | True   | admin        |
| 58f4decf0640fd24a8fde771 | observer     | True   | observer     |
| 58f4decf0640fd24a8fde772 | system_admin | True   | system_admin |
+--------------------------+--------------+--------+--------------+
```

### st2 role-assignment CLI commands

```bash
st2 role-assignment list
+--------------------------+----------+-------+-----------+----------------------------------+
| id                       | role     | user  | is_remote | description                      |
+--------------------------+----------+-------+-----------+----------------------------------+
| 58f4e3ad0640fd2a7948632d | observer | tomaz | False     |                                  |
| 59045aef58a5f53a3c85d417 | x        | a     | False     |                                  |
| 59045bed58a5f53a3c85d418 | x        | tomaz | False     |                                  |
| 5909c9140640fd0ee1b3981f | role_one | tomaz | True      | Automatic role assignment based  |
|                          |          |       |           | on the remote user membership in |
|                          |          |       |           | group "CN=stormers,OU=groups,DC= |
|                          |          |       |           | stackstorm,DC=net"               |
| 5909c9140640fd0ee1b39820 | role_two | tomaz | True      | Automatic role assignment based  |
|                          |          |       |           | on the remote user membership in |
|                          |          |       |           | group "CN=stormers,OU=groups,DC= |
|                          |          |       |           | stackstorm,DC=net"               |
+--------------------------+----------+-------+-----------+----------------------------------+
```

```bash
st2 role-assignment list --user=a
+--------------------------+------+------+-----------+-------------+
| id                       | role | user | is_remote | description |
+--------------------------+------+------+-----------+-------------+
| 59045aef58a5f53a3c85d417 | x    | a    | False     |             |
+--------------------------+------+------+-----------+-------------+
```

## TODO

- [x] RBAC for role assignments API endpoint (admins can view all the assignments, users can view their own assignments)
- [x] Tests for role assignments API RBAC

### CODE REVIEW

**Comment by LindsayHill** [comment]
Cool - new commands look really useful

File: st2common/st2common/openapi.yaml [hunk 1]
```diff
@@ -2069,6 +2074,72 @@ paths:
           description: Unexpected error
           schema:
             $ref: '#/definitions/Error'
+  /api/v1/rbac/role_assignments:
```
Comment (Mierdin) [inline st2common/st2common/openapi.yaml]: How are we handling API versioning? This seems like a substantial change; even though it's purely additive, I'm wondering when we should worry about bumping versions.

File: st2client/st2client/commands/rbac.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+                                       'output. "all" will return all '
+                                       'attributes.'))
+        self.parser.add_argument('-w', '--width', nargs='+', type=int,
+                                 default=None,
+                                 help=('Set the width of columns in output.'))
+
+    @resource.add_auth_token_to_kwargs_from_cli
+    def run(self, args, **kwargs):
+        # Filtering options
+        if args.system:
```
Comment (Mierdin) [inline st2client/st2client/commands/rbac.py]: What's this for? Seems strange.

File: st2common/st2common/models/api/rbac.py [hunk 3]
```diff
@@ -31,7 +34,7 @@
 
 
 class RoleAPI(BaseAPI):
-    model = PackDB
```
Comment (Mierdin) [inline st2common/st2common/models/api/rbac.py]: How this this work before? Doesn't seem to me that the PackDB model resembles roles at all

**Comment by Mierdin** [comment]
Also remember to merge the latest master into this now that we've fixed the mongoengine dep issue, to make sure u16 passes

File: st2common/st2common/openapi.yaml (hunk 1 above)
Comment (Kami) [inline st2common/st2common/openapi.yaml]: Yeah, for now we are just adding "additive" non-breaking changes to v1 and we don't really have other versioning scheme in place.

If and when we start changing API responses for existing endpoints, we will probably move to v2 or similar.

Oh, and we add "experimental" non-stable stuff to /exp/.

File: st2client/st2client/commands/rbac.py (hunk 2 above)
Comment (Kami) [inline st2client/st2client/commands/rbac.py]: It's a boolean filter which allows user to only display system roles.

`if args.system is None` would also work and perhaps be better, but right now, we only have positive flag and filter in most places so I went with that.

If we also wanted to support inverse operation (only display non-system roles), I would need to change it to `is None` check (and also change how flags work).

File: st2common/st2common/models/api/rbac.py (hunk 3 above)
Comment (Kami) [inline st2common/st2common/models/api/rbac.py]: This code path and model attribute is only used when creating objects through the API and that's not what happens with RBAC (we have a read-only API) so this code path was never exercised.

**Comment by Kami** [comment]
@Mierdin Done - thanks for the reminder.

**Pull request merged**